### PR TITLE
CD: Add env variables to application [Part 2]

### PR DESCRIPTION
**Pull request by SarahJaine**
**Description of change**
In https://github.com/adhocteam/Head-Start-TTADP/pull/29 I added the env variables to the application.  However, the values were interpreted as string literals (not as the variable values); the deployed application had values like `AUTH_CLIENT_ID: ${SANDBOX_AUTH_CLIENT_ID}`.  

~Because circle ci doesn't read/interpret the deployment config files and transform the circleci project env variables, it's necessary to write the true values in "manually." (I used `sed`.)  I'm not exactly thrilled with this approach- as it leaves plain text secrets in the deployment env. The files/secrets ARE NOT copied onto the application instance though, because the deployment directory is included in .cfignore.~

~I tried two other strategies that proved unsuccessful.~
~1. Using `cf set-env` to pass env variables: This doesn't work well for us because it can only be used on a running application, then the application must be restaged for variables to take hold.  Given that nearly every sandbox deploy is not to a running application, prod has never been deployed, and I seem to be killed dev often, this doesn't seem like a good strategy for us.  I could always push, set-env, then restage but that isn't very efficient.~
~2. Using `cf push` flag `--var`: After a few hours of guess-and-check, I could not get the syntax of this to work. I can't find ANY examples on the internet of people using this to set environment variables (things like `instance` count and app `name` yes, but not true env variables.)~

@rahearn clued me in on how to correctly user the `--var` flag on `cf push` (strategy 2 in the above comments).  Variable substitution now works as follows. 
1. A secret is set in the circle ci environment variable list. `SANDBOX_AUTH_CLIENT_ID: secret`
2. The deployment pipeline invokes the `cf_depoy` command with the parameter `auth_client_id: SANDBOX_AUTH_CLIENT_ID`. 
3. Within the `cf deploy` command, `cf push` is called with the vars-file where `AUTH_CLIENT_ID: ((AUTH_CLIENT_ID))`.  On the same line, `--var` sets `((AUTH_CLIENT_ID))` equal to the `SANDBOX_AUTH_CLIENT_ID` circle ci environment variable.  The application manifest interprets all this and sets `AUTH_CLIENT_ID` equal to the `SANDBOX_AUTH_CLIENT_ID`.

This is better because secrets aren't being written to deployment config files. Before, .cfignore was the only thing keeping those secret laden deployment config files off of the application instance. This implementation keeps env variables where they should be, in the environment.

**How to test**
- check that most recent deployment on this branch passed https://app.circleci.com/pipelines/github/adhocteam/Head-Start-TTADP?branch=sj-fix-env-var
- check that variables are passed into sandbox application instance https://dashboard.fr.cloud.gov/applications/2oBn9LBurIXUNpfmtZCQTCHnxUM/150aceb1-bcd6-4a5d-961d-a8d8e1c1f1b5/variables
- check that sandbox application is running https://tta-smarthub-sandbox.app.cloud.gov/. After login, the redirect will fail with a 4xx (instead of it's previous 5xx error) until 

**Issue(s)**
* https://github.com/HHS/Head-Start-TTADP/issues/47

**Checklist**

- [x] Meets issue criteria
- [x] Code tested
- [n/a] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [x] Documentation updated - **_Added new readme to deployment vars dir, added section on secret management to main readme_**
    - API methods
    - Boundary and Data Flow Diagrams
    - [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions with the [Nygard template](https://github.com/joelparkerhenderson/architecture_decision_record/blob/master/adr_template_by_michael_nygard.md)
    - OSCAL templates completed when security controls are implemented or modified

### CODE REVIEW

**Review by jasalisbury — APPROVED** [review]

File: deployment_config/dev_vars.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
-AUTH_CLIENT_ID: ${AUTH_CLIENT_ID}
-AUTH_CLIENT_SECRET: ${AUTH_CLIENT_SECRET}
+AUTH_CLIENT_ID:
```
[the remark this answers is not in the document]
Comment (SarahJaine) [reply deployment_config/dev_vars.yml]: Good point-- in two months I will look at this and have no memory of how any of this works. I added a README to this directory to try to explain how these files set deployment variables. I also added a secret management section to the main readme.  Do you think that would help you get started if you needed to make changes in the future and I wasn't around?

**Review by jasalisbury — COMMENTED** [review]
So if I need to add a new secret I would:
 * Add the secret to circle CI (via the dashboard?)
 * Add the secret (with a name that matches the secret added above) to the parameter list, push command and the deploy job in `.circleci/conifg.yml`
 * Add the secret (with matching name) to the three `deployment_config/*` files

File: deployment_config/sandbox_vars.yml [hunk 2]
```diff
@@ -1,6 +1,6 @@
-AUTH_CLIENT_ID: ${AUTH_CLIENT_ID}
-AUTH_CLIENT_SECRET: ${AUTH_CLIENT_SECRET}
+AUTH_CLIENT_ID: ((AUTH_CLIENT_ID))
```
Comment (SarahJaine) [inline deployment_config/sandbox_vars.yml:1]: @rahearn Do you mind sanity checking this? The deployment appears to work.. and the variables make it into cloud.gov correctly.  It's feeling a bit ♾️  though.

**Comment by SarahJaine** [comment]
> So if I need to add a new secret I would:
> 
> * Add the secret to circle CI (via the dashboard?)
 Yes, you'd add it at 
> * Add the secret (with a name that matches the secret added above) to the parameter list, push command and the deploy job in `.circleci/conifg.yml`
Yes, something like `new_var: SANDBOX_NEW_VAR`, to the cf_deploy parameter list
> * Add the secret (with matching name) to the three `deployment_config/*` files

There's a little more than what you have there.

Say you wanted a new environment variable `MY_VAR` in the deployed application in the DEV environment.

1. Add true variable value to circle ci via the [env var dashboard](https://app.circleci.com/settings/project/github/adhocteam/Head-Start-TTADP/environment-variables?return-to=https%3A%2F%2Fcircleci.com%2Fdashboard), ex: `DEV_MY_VAR: secret`
2. Update deploy job where the `cf_deploy` command is invoked for the `dev` env conditional. You'll add a new line to include a new parameter.
```
- when:  # dev
          condition:
            and:
              - equal: [<< pipeline.project.git_url >>, << pipeline.parameters.dev_git_url >>]
              - equal: [<< pipeline.git.branch >>, << pipeline.parameters.dev_git_branch >>]
          steps:
            - cf_deploy:
                auth_client_id: DEV_AUTH_CLIENT_ID
                ...
                my_var: DEV_MY_VAR
```
3. Update `cf_deploy` command parameters list to expect `my_var`.
```
  cf_deploy:
    ...
    parameters:
      my_var:
        type: env_var_name
```
4. Amend `cf_deploy`'s `cf push` command to include the new variable with it's value assigned to the new parameter `--var MY_VAR=${<< parameters.my_var}`
5. Add `MY_VAR: ((MY_VAR))` to deployment_config/dev_vars.yml
6. In manifest.yml, under the `env` object add new variable.
```
applications:
- name: tta-smarthub-((env))
  ...
  env:
    MY_VAR: ((MY_VAR))
```

I was hoping step 5 wasn't necessary but it appears to be.

**Review by jasalisbury — APPROVED** [review]

File: deployment_config/sandbox_vars.yml (hunk 2 above)
Comment (SarahJaine) [reply deployment_config/sandbox_vars.yml:2]: You know what, nvm, I'm bolding soldiering on. :) I'll get your feedback on the big AH => HHS merge.

**Pull request merged**